File: BOJ/25_최단경로/1956_운동.py
import sys
import heapq

read = sys.stdin.readline
INF = int(1e9)

# 정점의 개수가 적고, 간선의 길이가 많기 때문에 플로이드 와샬이 시간적으로 빠르다.

def floyd():
    for k in range(1,V+1) :
        for i in range(1,V+1) :
            for j in range(1,V+1) :
                if adj[i][j] > adj[i][k] + adj[k][j] :
                    adj[i][j] = adj[i][k] + adj[k][j]
        
V,E = map(int,read().split())
adj = [[INF] * (V+1) for _ in range(V+1)]
for _ in range(E):
    x,y,c = map(int,read().split())
    adj[x][y] = c

floyd()
ans = INF
for i in range(1,V+1) :
    for j in range(1,V+1) :
        if i == j :
            ans = min(ans,adj[i][j])
if ans == INF :
    print(-1)
else :
    print(ans)

File: BOJ/25_최단경로/1956_운동.cpp
#include <vector>
#include <iostream>
#include <algorithm>
using namespace std;


#define MAXN 400
#define INF 987654321

vector<vector<int> > adj(MAXN+1, vector<int>(MAXN+1,INF));
int V,E,x,y,c;

void floyd() {
    for(int k=1; k<=V; k++) {
        for(int i=1; i<=V; i++) {
            for(int j=1; j<=V; j++){
                if(adj[i][j] > adj[i][k] + adj[k][j]){
                    adj[i][j] = adj[i][k] + adj[k][j];
                }
            }
        }
    }
}

int main() {
    
    scanf("%d %d",&V,&E);
    for(int i=0; i<E; i++) {
        scanf("%d %d %d",&x,&y,&c);
        adj[x][y] = c;
    }
    floyd();

    int ans = INF;
    for(int i=1; i<=V; i++) {
        for(int j=1; j<=V; j++){
            if(i == j) {
                ans = min(ans,adj[i][j]);
            }
        }
    }
    if(ans == INF) printf("%d\n",-1);
    else printf("%d\n",ans);
    return 0;
}